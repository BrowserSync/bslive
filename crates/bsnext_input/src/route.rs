use std::collections::BTreeMap;
use std::fmt::{Display, Formatter};
use std::hash::{Hash, Hasher};
use std::ops::Deref;

use crate::watch_opts::WatchOpts;
use bsnext_resp::inject_opts::InjectOpts;

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct Route {
    pub path: String,
    #[serde(flatten)]
    pub kind: RouteKind,
    #[serde(flatten)]
    pub opts: Opts,
}

#[derive(Debug, Default, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct Opts {
    #[serde(flatten)]
    pub cors: Option<CorsOpts>,
    #[serde(flatten)]
    pub delay: Option<DelayOpts>,
    #[serde(default)]
    pub watch: WatchOpts,
    #[serde(default)]
    pub inject: InjectOpts,
    pub headers: Option<BTreeMap<String, String>>,
}

impl Default for Route {
    fn default() -> Self {
        Self {
            path: "/".to_string(),
            kind: RouteKind::new_html("default"),
            opts: Opts {
                headers: None,
                cors: None,
                delay: None,
                watch: Default::default(),
                inject: Default::default(),
            },
        }
    }
}

impl AsRef<Route> for Route {
    fn as_ref(&self) -> &Route {
        self
    }
}

impl Route {
    pub fn path(&self) -> &str {
        self.path.as_str()
    }
}

#[derive(Debug, Hash, PartialEq, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum RouteKind {
    Raw(RawRoute),
    Proxy(ProxyRoute),
    Dir(DirRoute),
}

impl RouteKind {
    pub fn new_html(html: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Html { html: html.into() })
    }
    pub fn new_raw(raw: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Raw { raw: raw.into() })
    }
    pub fn new_json(json: JsonWrapper) -> Self {
        RouteKind::Raw(RawRoute::Json { json })
    }
    pub fn new_sse(raw: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Sse { sse: raw.into() })
    }
}

#[derive(Debug, PartialEq, Clone, serde::Deserialize, serde::Serialize)]
pub struct JsonWrapper(pub serde_json::Value);

impl Display for JsonWrapper {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", serde_json::to_string(&self.0).expect("serde_json"))
    }
}

impl Deref for JsonWrapper {
    type Target = serde_json::Value;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Hash for JsonWrapper {
    fn hash<H: Hasher>(&self, state: &mut H) {
        //todo: implement hashing for serde_value
        if let Ok(as_str) = serde_json::to_string(&self.0) {
            state.write(as_str.as_bytes());
        } else {
            todo!("handle error here?")
        }
    }
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct DirRoute {
    pub dir: String,
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct ProxyRoute {
    pub proxy: String,
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum RawRoute {
    Html { html: String },
    Json { json: JsonWrapper },
    Raw { raw: String },
    Sse { sse: String },
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum CorsOpts {
    #[serde(rename = "cors")]
    Cors(bool),
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum DelayOpts {
    #[serde(rename = "delay")]
    Delay(DelayKind),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub enum DelayKind {
    #[serde(rename = "ms")]
    Ms(u64),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub enum DebounceDuration {
    #[serde(rename = "ms")]
    Ms(u64),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
#[serde(untagged)]
pub enum FilterKind {
    StringGlob(String),
    Extension { ext: String },
    Glob { glob: String },
    List(Vec<FilterKind>),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct Spec {
    #[serde(flatten)]
    pub opts: Option<SpecOpts>,
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct SpecOpts {
    pub debounce: Option<DebounceDuration>,
    pub filter: Option<FilterKind>,
}

#[derive(
    Debug, PartialOrd, Ord, Eq, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct Watcher {
    pub dir: String,
    pub debounce_ms: Option<u64>,
    pub filter: Option<FilterKind>,
}
