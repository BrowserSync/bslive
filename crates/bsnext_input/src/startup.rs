use crate::{Input, InputError};
use std::env::current_dir;
use std::path::PathBuf;

pub struct Startup {
    pub tasks: Vec<StartupTask>,
}

#[derive(Debug)]
pub struct StartupContext {
    pub cwd: PathBuf,
}

impl StartupContext {
    pub fn from_cwd(cwd: Option<&PathBuf>) -> Self {
        StartupContext {
            cwd: cwd.map(ToOwned::to_owned).unwrap_or_else(|| {
                PathBuf::from(
                    current_dir()
                        .expect("if current_dir fails, nothing can work")
                        .to_string_lossy()
                        .to_string(),
                )
            }),
        }
    }
}

#[derive(Debug)]
pub enum SystemStartArgs {
    PathWithInput {
        path: PathBuf,
        input: Input,
    },
    InputOnly {
        input: Input,
    },
    PathWithInvalidInput {
        path: PathBuf,
        input_error: InputError,
    },
}

pub trait SystemStart {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, Box<InputError>>;
}

impl Default for StartupContext {
    fn default() -> Self {
        Self::from_cwd(None)
    }
}

pub enum StartupTask {}
