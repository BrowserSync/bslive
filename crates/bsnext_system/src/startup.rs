use bsnext_input::{Input, InputError};
use std::env::current_dir;
use std::path::PathBuf;

pub type StartupResult = Result<DidStart, StartupError>;

pub struct Startup {
    pub tasks: Vec<StartupTask>,
}
#[derive(Debug)]
pub struct StartupContext {
    pub cwd: PathBuf,
}

impl StartupContext {
    pub fn from_cwd(cwd: Option<&PathBuf>) -> Self {
        StartupContext {
            cwd: cwd.map(ToOwned::to_owned).unwrap_or_else(|| {
                PathBuf::from(
                    current_dir()
                        .expect("if current_dir fails, nothing can work")
                        .to_string_lossy()
                        .to_string(),
                )
            }),
        }
    }
}

pub trait SystemStart {
    fn input(&self, ctx: &StartupContext) -> Result<(Input, Option<PathBuf>), InputError>;
}

impl Default for StartupContext {
    fn default() -> Self {
        Self::from_cwd(None)
    }
}

#[derive(Debug)]
pub enum DidStart {
    Started,
}

#[derive(Debug, thiserror::Error)]
pub enum StartupError {
    #[error("An input error prevented startup")]
    InputError(#[from] InputError),
}

pub enum StartupTask {}
