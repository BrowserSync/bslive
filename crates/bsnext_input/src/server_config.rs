use crate::client_config::ClientConfig;
use crate::playground::Playground;
use crate::route::{Route, Watcher};
use crate::{rand_word, PortError};
use serde::{de, Deserializer};
use std::hash::{DefaultHasher, Hash, Hasher};
use std::net::SocketAddr;
use std::str::FromStr;

#[derive(Debug, Default, Clone, serde::Deserialize, serde::Serialize)]
pub struct ServerConfig {
    #[serde(flatten)]
    pub identity: ServerIdentity,
    #[serde(default)]
    pub routes: Vec<Route>,
    #[serde(default)]
    pub watchers: Vec<Watcher>,
    #[serde(default)]
    pub playground: Option<Playground>,
    #[serde(default)]
    pub clients: ClientConfig,
}

impl ServerConfig {
    ///
    /// All regular routes, plus dynamically added ones (for example, through a playground)
    ///
    pub fn combined_routes(&self) -> Vec<Route> {
        let routes = self.routes.clone();
        match &self.playground {
            None => self.routes.clone(),
            Some(playground) => {
                let mut pg_routes = playground.as_routes();
                pg_routes.extend(routes);
                pg_routes
            }
        }
    }
    pub fn raw_routes(&self) -> &[Route] {
        &self.routes
    }

    pub fn from_route(r: Route) -> Self {
        Self {
            routes: vec![r],
            ..std::default::Default::default()
        }
    }
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Hash, Eq, Clone, serde::Deserialize, serde::Serialize,
)]
#[serde(untagged)]
pub enum ServerIdentity {
    PortNamed {
        #[serde(deserialize_with = "deserialize_port")]
        port: u16,
        name: String,
    },
    Both {
        name: String,
        bind_address: String,
    },
    Address {
        bind_address: String,
    },
    Named {
        name: String,
    },
    Port {
        #[serde(deserialize_with = "deserialize_port")]
        port: u16,
    },
}

fn deserialize_port<'de, D>(deserializer: D) -> Result<u16, D::Error>
where
    D: Deserializer<'de>,
{
    struct PortVisitor;

    impl<'de> de::Visitor<'de> for PortVisitor {
        type Value = u16;

        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
            formatter.write_str("a string or integer representing a valid port number")
        }

        fn visit_u64<E>(self, value: u64) -> Result<u16, E>
        where
            E: de::Error,
        {
            if value <= u16::MAX as u64 {
                Ok(value as u16)
            } else {
                Err(E::custom(format!("port number out of range: {}", value)))
            }
        }

        fn visit_str<E>(self, value: &str) -> Result<u16, E>
        where
            E: de::Error,
        {
            value
                .parse::<u16>()
                .map_err(|_| E::custom(format!("invalid port number: {}", value)))
        }
    }

    deserializer.deserialize_any(PortVisitor)
}

impl Default for ServerIdentity {
    fn default() -> Self {
        Self::named()
    }
}

impl ServerIdentity {
    pub fn named() -> Self {
        Self::Named { name: rand_word() }
    }
    pub fn address<A: AsRef<str>>(a: A) -> Self {
        Self::Address {
            bind_address: a.as_ref().to_string(),
        }
    }
    /// if a port was provided, try to use it by validating it first
    /// otherwise default to a named identity
    pub fn from_port_or_named(port: Option<u16>) -> Result<Self, PortError> {
        let result = port.map(|port| {
            SocketAddr::from_str(&format!("0.0.0.0:{port}"))
                .map_err(|err| PortError::InvalidPort { port, err })
        });
        match result {
            None => Ok(ServerIdentity::named()),
            Some(Ok(addr)) => Ok(ServerIdentity::address(addr.to_string())),
            Some(Err(err)) => Err(err),
        }
    }

    pub fn as_id(&self) -> u64 {
        let mut hasher = DefaultHasher::new();
        self.hash(&mut hasher);
        hasher.finish()
    }

    pub fn is_named(&self, pred_name: &str) -> bool {
        match self {
            ServerIdentity::Both { name, .. } => name == pred_name,
            ServerIdentity::Address { .. } => false,
            ServerIdentity::Named { name } => name == pred_name,
            ServerIdentity::Port { .. } => false,
            ServerIdentity::PortNamed { name, .. } => name == pred_name,
        }
    }
}

#[test]
fn server_config_as_enum() {
    #[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
    struct C {
        servers: Vec<ServerConfig>,
    }
    let input = r#"
servers:
    - bind_address: 127.0.0.1:3000
    - name: server_1
    - name: server_2
      bind_address: 127.0.0.1:3001
      routes:
        - path: /
          dir: .
    "#;
    let c: C = serde_yaml::from_str(input).unwrap();

    // let baseline = ServerConfigInput::Named { name: "server_1".into() };
    let baseline = ServerIdentity::Address {
        bind_address: "127.0.0.1x:3000".into(),
    };
    for x in &c.servers {
        assert_eq!(x.identity == baseline, false);
    }
}

#[test]
fn with_port() {
    #[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
    struct C {
        servers: Vec<ServerConfig>,
    }
    let input = r#"
servers:
    - port: 3000
    - port: "3001"
    "#;
    let c: C = serde_yaml::from_str(input).unwrap();
    let expected = ServerIdentity::Port { port: 3000 };
    let first = c.servers.first().unwrap();
    assert_eq!(first.identity, expected);

    let second = c.servers.get(1).unwrap();
    let second_expected = ServerIdentity::Port { port: 3001 };
    assert_eq!(second.identity, second_expected);
}
