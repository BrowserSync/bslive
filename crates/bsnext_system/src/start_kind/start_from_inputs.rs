use crate::startup::{StartupContext, SystemStart, SystemStartArgs};

use bsnext_input::{Input, InputError};
use std::path::{Path, PathBuf};

#[derive(Debug)]
pub struct StartFromInputPaths {
    pub input_paths: Vec<String>,
}

impl SystemStart for StartFromInputPaths {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, InputError> {
        from_yml_paths(&ctx.cwd, &self.input_paths)
    }
}

fn from_yml_paths<T: AsRef<str>>(cwd: &Path, inputs: &[T]) -> Result<SystemStartArgs, InputError> {
    let input_candidates = inputs
        .iter()
        .map(|path| cwd.join(path.as_ref()))
        .collect::<Vec<PathBuf>>();

    let lookups = ["input.yml", "input.yaml"]
        .iter()
        .map(|path| cwd.join(path))
        .collect::<Vec<PathBuf>>();

    let auto_candidates = lookups
        .iter()
        .filter(|path| Path::exists(path))
        .collect::<Vec<&PathBuf>>();

    let exists = input_candidates
        .iter()
        .filter(|path| Path::exists(path))
        .collect::<Vec<&PathBuf>>();

    let missing = input_candidates
        .iter()
        .filter(|path| !Path::exists(path))
        .collect::<Vec<&PathBuf>>();

    if !missing.is_empty() {
        for path in &missing {
            tracing::error!(?path, "input file not found");
        }
        return Err(InputError::NotFound(
            missing.first().expect("guarded").to_path_buf(),
        ));
    }

    let first_user = exists.first();
    let first_auto = auto_candidates.first();

    let Some(input_path) = first_user.or(first_auto) else {
        return Err(InputError::MissingInputs);
    };

    tracing::info!(?input_path);

    let result = Input::from_input_path(input_path);
    match result {
        Ok(input) => Ok(SystemStartArgs::PathWithInput {
            path: input_path.to_path_buf(),
            input,
        }),
        Err(InputError::YamlError(yaml_error)) => Ok(SystemStartArgs::PathWithInvalidInput {
            path: input_path.to_path_buf(),
            input_error: InputError::YamlError(yaml_error),
        }),
        Err(e) => {
            tracing::error!("cannot continue");
            Err(e)
        }
    }
}
