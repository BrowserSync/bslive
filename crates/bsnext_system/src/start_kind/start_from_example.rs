use bsnext_example::Example;
use bsnext_input::server_config::Identity;
use bsnext_input::startup::{StartupContext, SystemStart, SystemStartArgs};
use bsnext_input::target::TargetKind;
use bsnext_input::{fs_write_input, rand_word, DirError, InputError};
use std::fs;

#[derive(Debug)]
pub struct StartFromExample {
    pub example: Example,
    pub write_input: bool,
    pub target_kind: TargetKind,
    pub port: Option<u16>,
    pub temp: bool,
    pub name: Option<String>,
}

impl SystemStart for StartFromExample {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, InputError> {
        let identity = Identity::from_port_or_named(self.port)?;
        let input = self.example.into_input(identity);
        let name = self.name.clone();
        let dir = if self.temp {
            let temp_dir = std::env::temp_dir();
            let word = name.unwrap_or_else(rand_word);
            let num = rand::random::<f64>();
            let next_dir = temp_dir.join(format!("bslive-{word}-{num}"));
            fs::create_dir_all(&next_dir)
                .map_err(|_e| DirError::CannotCreate {
                    path: next_dir.clone(),
                })
                .and_then(|_pb| {
                    std::env::set_current_dir(&next_dir).map_err(|_e| DirError::CannotMove {
                        path: next_dir.clone(),
                    })
                })
                .map(|_| next_dir.clone())?
        } else {
            ctx.cwd.to_path_buf()
        };
        if self.write_input {
            let path = fs_write_input(&dir, &input, self.target_kind.clone())?;
            Ok(SystemStartArgs::PathWithInput { path, input })
        } else {
            Ok(SystemStartArgs::InputOnly { input })
        }
    }
}
