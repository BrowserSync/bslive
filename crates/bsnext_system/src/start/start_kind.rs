use crate::start::start_command::StartCommand;
use crate::start::start_kind::start_from_example::StartFromExample;
use crate::start::start_kind::start_from_inputs::{StartFromInput, StartFromInputPaths};
use crate::start::start_kind::start_from_paths::StartFromTrailingArgs;
use bsnext_core::shared_args::{FsOpts, InputOpts};
use bsnext_fs_helpers::{fs_write_str, FsWriteError, WriteMode};
use bsnext_input::route::{CorsOpts, Opts, Route};
use bsnext_input::server_config::{ServerConfig, ServerIdentity};
use bsnext_input::startup::{StartupContext, SystemStart, SystemStartArgs};
use bsnext_input::target::TargetKind;
use bsnext_input::InputWriter;
use bsnext_input::{Input, InputError};
use std::path::{Path, PathBuf};

pub mod start_from_example;
pub mod start_from_inputs;
pub mod start_from_paths;

#[allow(clippy::large_enum_variant)]
#[derive(Debug)]
pub enum StartKind {
    FromInput(StartFromInput),
    FromInputPaths(StartFromInputPaths),
    FromExample(StartFromExample),
    FromTrailingArgs(StartFromTrailingArgs),
}

impl StartKind {
    pub fn from_args(fs_opts: &FsOpts, input_opts: &InputOpts, cmd: &StartCommand) -> Self {
        // todo: re-implement example command
        // if let Some(example) = args.example {
        //     return StartKind::FromExample(StartFromExample {
        //         example,
        //         write_input: args.write,
        //         port: args.port,
        //         temp: args.temp,
        //         name: args.name.clone(),
        //         target_kind: args
        //             .target
        //             .as_ref()
        //             .map(ToOwned::to_owned)
        //             .unwrap_or_default(),
        //         dir: args.dir.clone(),
        //         force: args.force,
        //     });
        // }

        // todo: make the addition of a proxy + route opts easier?
        if cmd.trailing.is_empty() {
            tracing::debug!("0 trailing, {} inputs", input_opts.input.len());
            if input_opts.input.is_empty() && !cmd.proxies.is_empty() {
                let first_proxy = cmd.proxies.first().expect("guarded first proxy");
                let r = Route::proxy(first_proxy);
                let id = ServerIdentity::from_port_or_named(cmd.port).unwrap_or_else(|_e| {
                    tracing::error!("A problem occurred with the port?");
                    ServerIdentity::named()
                });
                let ser = ServerConfig::from_route(r, id);
                let input = Input::from_server(ser);
                StartKind::FromInput(StartFromInput { input })
            } else {
                StartKind::FromInputPaths(StartFromInputPaths {
                    input_paths: input_opts.input.clone(),
                    port: cmd.port,
                })
            }
        } else {
            StartKind::FromTrailingArgs(StartFromTrailingArgs {
                paths: cmd.trailing.clone(),
                write_input: fs_opts.write,
                port: cmd.port,
                force: fs_opts.force,
                route_opts: Opts {
                    cors: cmd.cors.then_some(CorsOpts::Cors(true)),
                    ..Default::default()
                },
            })
        }
    }
    pub fn from_input(input: Input) -> Self {
        Self::FromInput(StartFromInput { input })
    }
}

impl SystemStart for StartKind {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, Box<InputError>> {
        match self {
            StartKind::FromInputPaths(from_inputs) => from_inputs.input(ctx),
            StartKind::FromExample(from_example) => from_example.input(ctx),
            StartKind::FromTrailingArgs(from_trailing_args) => from_trailing_args.input(ctx),
            StartKind::FromInput(from_inputs) => from_inputs.input(ctx),
        }
    }
}

pub fn fs_write_input(
    cwd: &Path,
    input: &Input,
    target_kind: TargetKind,
    write_mode: &WriteMode,
) -> Result<PathBuf, FsWriteError> {
    let string = match target_kind {
        TargetKind::Yaml => bsnext_yaml::yaml_writer::YamlWriter.input_to_str(input),
        TargetKind::Toml => todo!("toml missing"),
        TargetKind::Md => bsnext_md::md_writer::MdWriter.input_to_str(input),
        TargetKind::Html => bsnext_html::html_writer::HtmlWriter.input_to_str(input),
    };
    let name = match target_kind {
        TargetKind::Yaml => "bslive.yml",
        TargetKind::Toml => todo!("toml missing"),
        TargetKind::Md => "bslive.md",
        TargetKind::Html => "bslive.html",
    };

    fs_write_str(cwd, &PathBuf::from(name), &string, write_mode)
}
