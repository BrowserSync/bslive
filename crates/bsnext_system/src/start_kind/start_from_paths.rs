use crate::start_kind::start_fs;
use bsnext_fs_helpers::WriteMode;
use bsnext_input::route::{DirRoute, Route, RouteKind};
use bsnext_input::server_config::{ServerConfig, ServerIdentity};
use bsnext_input::startup::{StartupContext, SystemStart, SystemStartArgs};
use bsnext_input::target::TargetKind;
use bsnext_input::{Input, InputError, PathDefinition, PathDefs, PathError};
use std::path::{Path, PathBuf};

#[derive(Debug)]
pub struct StartFromDirPaths {
    pub paths: Vec<String>,
    pub write_input: bool,
    pub port: Option<u16>,
    pub force: bool,
}

impl SystemStart for StartFromDirPaths {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, Box<InputError>> {
        let identity =
            ServerIdentity::from_port_or_named(self.port).map_err(|e| Box::new(e.into()))?;
        let input =
            from_dir_paths(&ctx.cwd, &self.paths, identity).map_err(|e| Box::new(e.into()))?;
        let write_mode = if self.force {
            WriteMode::Override
        } else {
            WriteMode::Safe
        };
        if self.write_input {
            let path = start_fs::fs_write_input(&ctx.cwd, &input, TargetKind::Yaml, &write_mode)
                .map_err(|e| Box::new(e.into()))?;
            Ok(SystemStartArgs::PathWithInput { input, path })
        } else {
            Ok(SystemStartArgs::InputOnly { input })
        }
    }
}

fn from_dir_paths<T: AsRef<str>>(
    cwd: &Path,
    paths: &[T],
    identity: ServerIdentity,
) -> Result<Input, PathError> {
    let path_defs = paths
        .iter()
        .map(|p| {
            let pb = PathBuf::from(p.as_ref());
            if pb.is_absolute() {
                return PathDefinition {
                    input: p.as_ref().to_string(),
                    cwd: cwd.to_path_buf(),
                    absolute: pb,
                };
            } else {
                return PathDefinition {
                    input: p.as_ref().to_string(),
                    cwd: cwd.to_path_buf(),
                    absolute: cwd.join(pb),
                };
            }
        })
        .map(|path_def| {
            let exists = path_def.absolute.exists();
            (path_def, exists)
        })
        .collect::<Vec<(PathDefinition, bool)>>();

    let invalid = path_defs
        .into_iter()
        .filter_map(|(pb, exists)| if exists { None } else { Some(pb) })
        .collect::<Vec<_>>();

    if !invalid.is_empty() {
        return Err(PathError::MissingPaths {
            paths: PathDefs(invalid),
        });
    }

    let server = ServerConfig {
        identity,
        routes: paths
            .iter()
            .map(|p| -> Route {
                let str = p.as_ref();
                Route {
                    path: "/".to_string().parse().unwrap(),
                    kind: RouteKind::Dir(DirRoute {
                        dir: str.into(),
                        base: None,
                    }),
                    ..Default::default()
                }
            })
            .collect(),
        ..Default::default()
    };
    Ok(Input {
        servers: vec![server],
    })
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn test() -> anyhow::Result<()> {
        use tempfile::tempdir;
        let tmp_dir = tempdir()?;
        let v = StartFromDirPaths {
            paths: vec![".".into()],
            write_input: false,
            port: Some(3000),
            force: false,
        };
        let ctx = StartupContext {
            cwd: tmp_dir.path().to_path_buf(),
        };
        let i = v.input(&ctx);
        tmp_dir.close()?;
        let start_args = i.unwrap();
        if let SystemStartArgs::InputOnly { input } = start_args {
            insta::assert_debug_snapshot!(input);
            insta::assert_yaml_snapshot!(input);
        } else {
            unreachable!("cannot get here?")
        }
        Ok(())
    }
}
