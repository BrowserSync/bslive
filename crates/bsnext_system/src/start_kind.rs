use crate::args::Args;
use crate::start_kind::start_from_example::StartFromExample;
use crate::start_kind::start_from_inputs::{StartFromInput, StartFromInputPaths};
use crate::start_kind::start_from_paths::StartFromPaths;
use bsnext_input::startup::{StartupContext, SystemStart, SystemStartArgs};

use bsnext_input::{Input, InputError};

pub mod start_from_example;
pub mod start_from_inputs;
pub mod start_from_paths;

#[derive(Debug)]
pub enum StartKind {
    FromInput(StartFromInput),
    FromInputPaths(StartFromInputPaths),
    FromExample(StartFromExample),
    FromPaths(StartFromPaths),
}

impl StartKind {
    pub fn from_args(args: Args) -> Self {
        if let Some(example) = args.example {
            return StartKind::FromExample(StartFromExample {
                example,
                write_input: args.write,
                port: args.port,
                temp: args.temp,
                name: args.name,
                target_kind: args.target.unwrap_or_default(),
            });
        }

        if !args.paths.is_empty() {
            StartKind::FromPaths(StartFromPaths {
                paths: args.paths,
                write_input: args.write,
                port: args.port,
            })
        } else {
            StartKind::FromInputPaths(StartFromInputPaths {
                input_paths: args.input.clone(),
            })
        }
    }
    pub fn from_input(input: Input) -> Self {
        Self::FromInput(StartFromInput { input })
    }
}

impl SystemStart for StartKind {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, InputError> {
        match self {
            StartKind::FromInputPaths(from_inputs) => from_inputs.input(ctx),
            StartKind::FromExample(from_example) => from_example.input(ctx),
            StartKind::FromPaths(from_input_paths) => from_input_paths.input(ctx),
            StartKind::FromInput(from_inputs) => from_inputs.input(ctx),
        }
    }
}
