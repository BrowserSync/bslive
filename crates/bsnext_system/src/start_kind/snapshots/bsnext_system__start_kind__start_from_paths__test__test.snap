---
source: crates/bsnext_system/src/start_kind/start_from_paths.rs
expression: input
---
Input {
    servers: [
        ServerConfig {
            identity: Address {
                bind_address: "0.0.0.0:3000",
            },
            routes: [
                Route {
                    path: "/",
                    kind: Dir(
                        DirRoute {
                            dir: ".",
                            base: None,
                        },
                    ),
                    opts: Opts {
                        cors: None,
                        delay: None,
                        watch: Bool(
                            true,
                        ),
                        inject: Bool(
                            true,
                        ),
                        headers: None,
                        compression: None,
                    },
                    fallback: None,
                },
            ],
            watchers: [],
            clients: ClientConfig {
                log: Info,
            },
        },
    ],
}
